idaes/apps/caprese/data_manager.py

```diff
@@ -0,0 +1,361 @@
+        plant_states_to_save = self.plant_user_interested_states + self.plantblock.differential_vars
+        self.plant_df = add_variable_values_to_dataframe(self.plant_df,
+                                                         plant_states_to_save, #no inputs
+                                                         iteration = 0,
+                                                         time_subset = [self.plantblock.time.first()])
+        
+
+    def save_plant_data(self, iteration):
+        #skip time.first()
+        time_subset = self.plantblock.time.ordered_data()[1:]
+        self.plant_df = add_variable_values_to_dataframe(self.plant_df,
+                                                         self.plant_vars_of_interest,
+                                                         iteration,
+                                                         time_subset = time_subset,)
+        
+class ControllerDataManager(PlantDataManager, NMPC_PlotLibrary):
+    def __init__(self, 
+                 plantblock, 
+                 controllerblock, 
+                 user_interested_states = [],
+                 user_interested_inputs = []):
+        
+        if not hasattr(self, "plant_df"):
+            super(ControllerDataManager, self).__init__(plantblock, 
+                                                        user_interested_states, 
+                                                        user_interested_inputs,)
+        
+        self.controllerblock = controllerblock
+        # Convert vars in plant to vars in controller
+        if user_interested_states != []:
+            cuid_states = [ComponentUID(var.referent) for var in user_interested_states]
+            self.controller_user_interested_states = [cuid.find_component_on(controllerblock)
+                                                      for cuid in cuid_states]
+        else:
+            self.controller_user_interested_states = []
+            
+        if user_interested_inputs != []:
+            cuid_inputs = [ComponentUID(var.referent) for var in user_interested_inputs]
+            self.controller_user_interested_inputs = [cuid.find_component_on(controllerblock)
+                                                      for cuid in cuid_inputs]
+        else:
+            self.controller_user_interested_inputs = []
+        
+        
+        self.controller_vars_of_interest = self.controller_user_interested_inputs + \
+                                                controllerblock.input_vars   
+        self.controller_df = empty_dataframe_from_variables(self.controller_vars_of_interest)
+        
+        self.states_need_setpoints = self.controller_user_interested_states + \
+                                        controllerblock.differential_vars
+        self.controller_df = add_setpoint_column_into_dataframe(self.controller_df, 
+                                                                self.states_need_setpoints)
```

Comment:
I save them in `controller_df` because I think only the controller knows the setpoint. But why store setpoints in a different dataframe? Let's talk about this.

Reply:
The reason would be so that we can use the same keys in both data frames. There is probably an argument for doing it either way.

Reply:
I think it is slightly cleaner to make the user know about the different dataframes then to make them know how we alter the variable's name in the dataframe key.